packages/checkout-storefront/src/sections/Addresses/BillingAddressSection.tsx

```diff
@@ -1,77 +1,148 @@
-import { useCheckoutAddressUpdate } from "./useCheckoutAddressUpdate";
+import { AddressFormData, UserAddressFormData } from "./types";
 import { UserAddressSection } from "./UserAddressSection";
+import { getAddressFormDataFromAddress, getAddressInputData, isMatchingAddress } from "./utils";
+import { getAddressVlidationRulesVariables } from "./utils";
 
-interface BillingAddressSectionProps {
-  addresses?: AddressFragment[] | null;
-  defaultBillingAddress: UserDefaultAddressFragment;
-}
-
-export const BillingAddressSection: React.FC<BillingAddressSectionProps> = ({
-  defaultBillingAddress,
-  addresses = [],
-}) => {
+export const BillingAddressSection = () => {
   const formatMessage = useFormattedMessages();
   const { user: authUser } = useAuthState();
   const { checkout } = useCheckout();
-  const {
-    isBillingSameAsShippingAddress,
-    hasBillingSameAsShippingAddressChanged,
-    setHasBillingSameAsShippingAddressChanged,
-  } = useBillingSameAsShipping();
+  const { billingAddress, shippingAddress, id: checkoutId } = checkout;
+
+  const hasBillingSameAsShipping = isMatchingAddress(shippingAddress, billingAddress);
+
+  const [useBillingSameAsShipping, setUseBillingSameAsShipping] = useState<boolean>(
+    !billingAddress || hasBillingSameAsShipping
+  );
+
+  const [{ data }] = useUserQuery({
+    pause: !authUser?.id,
+  });
+
+  const user = data?.me;
+  const addresses = user?.addresses;
+  const defaultBillingAddress = user?.defaultBillingAddress;
+  const errorProps = useErrors<AddressFormData>();
+  const { setApiErrors } = errorProps;
+  const [passDefaultFormDataAddress, setPassDefaultFormDataAddress] = useState<boolean>(
+    !!billingAddress
+  );
+
+  const defaultAddress = defaultBillingAddress;
+
+  const { showErrors } = useAlerts();
+
+  const [, checkoutBillingAddressUpdate] = useCheckoutBillingAddressUpdateMutation();
+  const useBillingSameAsShippingRef = useRef<boolean>(useBillingSameAsShipping);
+  const shippingAddressRef = useRef<AddressFragment | null | undefined>(shippingAddress);
+
+  const updateBillingAddress = async ({ autoSave, ...addressInput }: AddressFormData) => {
+    const result = await checkoutBillingAddressUpdate({
+      checkoutId,
+      billingAddress: getAddressInputData(addressInput),
+      validationRules: getAddressVlidationRulesVariables(autoSave),
+    });
 
-  const { setCountryCodeFromAddress } = useCountrySelect();
+    const [hasErrors, errors] = extractMutationErrors(result);
 
-  const { updateBillingAddress, billingErrorProps } = useCheckoutAddressUpdate();
+    if (hasErrors) {
+      showErrors(errors, "checkoutBillingUpdate");
+      setApiErrors(errors);
+    }
+  };
+
+  const setBillingSameAsShipping = async () => {
+    if (!hasBillingSameAsShipping) {
+      await updateBillingAddress({
+        ...getAddressFormDataFromAddress(shippingAddress),
+        autoSave: true,
+      } as AddressFormData);
```

Comment:
You're right. Unfortunately some of those are because api isn't typed best at the moment, for instance:
``` checkout?.shippingAddress?.country.code as CountryCode``` - the task for core has been waiting for some time but since it's a breaking change we have to wait till 4.0 
Talking to @korycins about making `user.addresses` an empty array by default so that should allow us to remove these:
```addresses={addresses as AddressFragment[]}```
Also went through the code and managed to type some stuff better so removed some of the assertions. Most of those that are left now are the two I mentioned above